modules/core/shared/src/main/scala/io/circe/JsonObject.scala

```diff
@@ -253,19 +252,20 @@ object JsonObject {
    * Construct a [[JsonObject]] from an [[scala.collection.Iterable]] (provided for optimization).
    */
   final def fromIterable(fields: Iterable[(String, Json)]): JsonObject = {
-    if (fields.isEmpty) {
-      empty
-    } else {
-      val map = new LinkedHashMap[String, Json]
-      val iterator = fields.iterator
-
-      while (iterator.hasNext) {
-        val (key, value) = iterator.next()
-
-        map.put(key, value)
-      }
-
-      fromLinkedHashMap(map)
+    val fieldsSeq = fields.iterator.toSeq
+    (fieldsSeq.size: @switch) match {
+      case 0 => empty
```

Comment:
`empty` is actually sub-optimial as well, maybe we could create a dedicated `empty` type as well instead of it being `final val empty: JsonObject = new MapAndVectorJsonObject(Map.empty, Vector.empty)`

Reply:
ok, I added one